bluemira/optimisation/problem.py

```diff
@@ -0,0 +1,123 @@
+# bluemira is an integrated inter-disciplinary design tool for future fusion
+# reactors. It incorporates several modules, some of which rely on other
+# codes, to carry out a range of typical conceptual fusion reactor design
+# activities.
+#
+# Copyright (C) 2021-2023 M. Coleman, J. Cook, F. Franza, I.A. Maione, S. McIntosh,
+#                         J. Morris, D. Short
+#
+# bluemira is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# bluemira is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+"""Interface for defining an optimisation problem."""
+import abc
+from typing import Any, Callable, List, Mapping, Optional, Tuple, TypeVar, Union
+
+import numpy as np
+
+from bluemira.optimisation._algorithm import Algorithm
+from bluemira.optimisation._optimise import OptimiserResult, optimise
+from bluemira.optimisation.typing import ConstraintT
+
+
+class OptimisationProblem(abc.ABC):
+    """
+    Interface for an optimisation problem.
+
+    This is an alternative to running an optimisation using the
+    :func:`.optimise` function.
+
+    Using this interface to define an optimisation can provide a few
+    benefits, including:
+
+        * Shared state between optimisation functions and constraints.
+          This can enable things like shared parameters and dynamic
+          constraints.
+        * Switch out optimisation problems using Liskov Substitution.
+        * Logical grouping of related functions.
+    """
+
+    @abc.abstractmethod
+    def objective(self, x: np.ndarray) -> float:
+        """The objective function to minimise."""
+
+    def df_objective(self, x: np.ndarray) -> np.ndarray:
+        """The gradient of the objective function at ``x``."""
+
+    def eq_constraints(self) -> List[ConstraintT]:
+        """The equality constraints on the optimisation."""
+        return []
+
+    def ineq_constraints(self) -> List[ConstraintT]:
+        """The inequality constraints on the optimisation."""
+        return []
+
+    def bounds(self) -> Optional[Tuple[np.ndarray, np.ndarray]]:
+        """The lower and upper bounds of the optimisation parameters."""
+        return None
```

Comment:
Should this not be an abstract method? Or can we mention what None does in terms of bounds? Does it default to -np.infty, np.infty` for example?

Reply:
I'll clarify - just trying to clean up how we handle bounds to make this easier. If we allow scalar bound inputs, we can expand those once we know the dimensionality of the problem, and we can be explicit here and return `(-np.inf, np.inf)`.